### PR TITLE
Allow column type mismatch between table and partition schema

### LINKED ISSUE #3035: Allow Hive partition type mismatch for same Presto types
We should allow reading from partitions that have different Hive types (such as `SMALLINT` and `INT`) that are implicitly coercible to the declared table type. For example, a table might have been created with type `SMALLINT` and have old partitions with that type, but the table type was later changed to `INT`.

### PR DESCRIPTION
**Pull request by geraint0923**
Fix #3035

### CODE REVIEW

**Comment by raghavsethi** [comment]
Is the GitHub ordering of commits the same as what you intend them to be reviewed in?

**Comment by geraint0923** [comment]
No, the last two are in reversed order, which means that the `Add test ...` one is the last commit.

File: presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java [hunk 1]
```diff
@@ -236,7 +237,8 @@ private void invokeFinishedIfNecessary()
                             files.getInputFormat(),
                             files.getSchema(),
                             files.getPartitionKeys(),
-                            files.getEffectivePredicate());
+                            files.getEffectivePredicate(),
+                            files.getMismatchColumnTypes());
```
Comment (raghavsethi) [inline presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java]: getColumnCoercions seems like a better name

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveSplitManager.java [hunk 2]
```diff
@@ -238,20 +243,26 @@ public ConnectorSplitSource getSplits(ConnectorTransactionHandle transaction, Co
...
-                                tableName,
-                                tableType,
-                                partName,
-                                partitionColumns.get(i).getName(),
-                                partitionType));
+                        if (coercionPolicy.canCoerce(partitionType, tableType)) {
+                            mismatchColumnTypes.put(i, partitionType);
+                        }
+                        else {
+                            throw new PrestoException(HIVE_PARTITION_SCHEMA_MISMATCH, format("" +
```
Comment (raghavsethi) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveSplitManager.java]: This error message needs to be more detailed. Something like - "there is a mismatch between table column type and partition column type, and coercion between these types is not permitted by policy".

**Comment by raghavsethi** [comment]
There are a couple of things that I'm concerned about:

There aren't enough tests - create a TestCoercionPolicy and verify that columns are actually read as different types. There should also be something in product tests that creates a table with a mismatched partition and reads it successfully.

Also, you have only implemented coercions for a single file format. This is fine, except that it behaves unpredictably. Users often have no idea what format they're reading, and can get different results based on this. I would put checks into all the other readers to fail if there are coercions required.

@martint would you agree with this assessment?

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveSplitManager.java (hunk 2 above)
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveSplitManager.java]: Also tell exactly which partition has he mismatch. This error is a top user question and thus needs to be as easy to understand as possible.

**Comment by geraint0923** [comment]
Superseded by #6255

**Pull request closed without merging**